### PR TITLE
Change CodeGenerateUtil so that it doesn't add/remove the code in an …

### LINKED ISSUE #11617: JPA Map Storage: Login fails when auth sessions are configured with CockroachDB
### Describe the bug

When auth sessions are configured to use the JPA Map Storage and CockroachDB is the database, the login process fails with

```
22:26:51,079 ERROR XNIO-1 task-2 [io.undertow.request] UT005023: Exception handling request to /auth/realms/master/login-actions/authenticate
org.keycloak.models.ModelException: javax.persistence.OptimisticLockException: org.hibernate.exception.LockAcquisitionException: could not execute statement
	at org.keycloak.connections.jpa.PersistenceExceptionConverter.convert(PersistenceExceptionConverter.java:84)
	at org.keycloak.models.map.storage.jpa.JpaTransactionWrapper.commit(JpaTransactionWrapper.java:53)
	at org.keycloak.services.DefaultKeycloakTransactionManager.commit(DefaultKeycloakTransactionManager.java:136)
	at org.keycloak.services.filters.AbstractRequestFilter.close(AbstractRequestFilter.java:64)
	at org.keycloak.services.filters.AbstractRequestFilter.filter(AbstractRequestFilter.java:49)
	at org.keycloak.testsuite.UndertowRequestFilter.doFilter(UndertowRequestFilter.java:47)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:68)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:117)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.SendErrorPageHandler.handleRequest(SendErrorPageHandler.java:52)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:269)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:78)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:133)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:130)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:249)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:78)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:99)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:387)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:841)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1982)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: javax.persistence.OptimisticLockException: org.hibernate.exception.LockAcquisitionException: could not execute statement
	at org.hibernate.internal.ExceptionConverterImpl.wrapLockException(ExceptionConverterImpl.java:277)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:98)
	at org.hibernate.internal.ExceptionConverterImpl.convert(ExceptionConverterImpl.java:181)
	at org.hibernate.internal.ExceptionConverterImpl.convertCommitException(ExceptionConverterImpl.java:65)
	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:107)
	at org.keycloak.models.map.storage.jpa.JpaTransactionWrapper.commit(JpaTransactionWrapper.java:51)
	... 38 more
Caused by: org.hibernate.exception.LockAcquisitionException: could not execute statement
	at org.hibernate.exception.internal.SQLStateConversionDelegate.convert(SQLStateConversionDelegate.java:123)
	at org.hibernate.exception.internal.StandardSQLExceptionConverter.convert(StandardSQLExceptionConverter.java:42)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:113)
	at org.hibernate.engine.jdbc.spi.SqlExceptionHelper.convert(SqlExceptionHelper.java:99)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:178)
	at org.hibernate.engine.jdbc.batch.internal.NonBatchingBatch.addToBatch(NonBatchingBatch.java:45)
	at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3499)
	at org.hibernate.persister.entity.AbstractEntityPersister.delete(AbstractEntityPersister.java:3756)
	at org.hibernate.action.internal.EntityDeleteAction.execute(EntityDeleteAction.java:99)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:604)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:478)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:356)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1472)
	at org.hibernate.internal.SessionImpl.managedFlush(SessionImpl.java:512)
	at org.hibernate.internal.SessionImpl.flushBeforeTransactionCompletion(SessionImpl.java:3310)
	at org.hibernate.internal.SessionImpl.beforeTransactionCompletion(SessionImpl.java:2506)
	at org.hibernate.engine.jdbc.internal.JdbcCoordinatorImpl.beforeTransactionCompletion(JdbcCoordinatorImpl.java:447)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.beforeCompletionCallback(JdbcResourceLocalTransactionCoordinatorImpl.java:178)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl.access$300(JdbcResourceLocalTransactionCoordinatorImpl.java:39)
	at org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImpl$TransactionDriverControlImpl.commit(JdbcResourceLocalTransactionCoordinatorImpl.java:271)
	at org.hibernate.engine.transaction.internal.TransactionImpl.commit(TransactionImpl.java:104)
	... 39 more
Caused by: org.postgresql.util.PSQLException: ERROR: restart transaction: TransactionRetryWithProtoRefreshError: TransactionRetryError: retry txn (RETRY_WRITE_TOO_OLD - WriteTooOld flag converted to WriteTooOldError): "sql txn" meta={id=49d8f12f key=/Table/377/2/"\x05\x15\xdaB\x1e\xa4L\uead1D\xfb\xbc\xdd}\xfa"/"\xa9!\x9f\x81\x05\xa7FN\x8f\xa6)\x82\uf21eP"/0 pri=0.00004675 epo=0 ts=1650677210.826817550,1 min=1650677210.770774855,0 seq=3} lock=true stat=PENDING rts=1650677210.770774855,0 wto=false gul=1650677211.270774855,0
  Hint: See: https://www.cockroachlabs.com/docs/v21.2/transaction-retry-error-reference.html#retry_write_too_old
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2675)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2365)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:355)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:490)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:408)
	at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:166)
	at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:134)
	at org.hibernate.engine.jdbc.internal.ResultSetReturnImpl.executeUpdate(ResultSetReturnImpl.java:175)
	... 56 more
```
The `kc_auth_session` and `kc_auth_root_session` tables are created and once you navigate to the login screen there's data that is added to the tables. When `Sign In` is clicked the exception above is shown and the data is not cleared as it happens on PostgreSQL.

### Version

18.0.0

### Expected behavior

Login should succeed

### Actual behavior

Login fails with the `RETRY_TOO_OLD` error message.

### How to Reproduce?

Enable the JPA map storage for auth sessions, and in the `jpa-map-storage` add a configuration that connects to a CockroachDB. Then either run any test that performs a login (e.g. `AccountBrokerTest`), or run `KeycloakServer`, go to the admin console and try to login.

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by sguilhen**
…inner transaction

Fixes #11617

### CODE REVIEW

**Comment by sguilhen** [comment]
To be able to test with CockroachDB, #12176 must be resolved first. Also, with this fix we have regressions:

```
[ERROR] Failures: 
[ERROR]   LDAPProvidersIntegrationTest.testImportExistingUserFromLDAP:737 expected:<[Username already exists].> but was:<[Action expired. Please continue with login now].>
[ERROR]   LDAPProvidersIntegrationNoImportTest>LDAPProvidersIntegrationTest.testImportExistingUserFromLDAP:737 expected:<[Username already exists].> but was:<[Action expired. Please continue with login now].>
```

**Comment by sguilhen** [comment]
**UPDATE:** The LDAP tests are passing with all storages now - this version that I've pushed was slightly different from the one I used to run the tests. With this version everything works.

So the only difference between CRDB and PostgreSQL at the moment, when it comes to the testsuite, is that the concurrent tests (ConcurrentLoginTests and the two concurrent test methods in ComponentTest) fail on CRDB. All other tests pass with both databases.

**Review by vramik — COMMENTED** [review]
Changes looks good to me.

**Review by vramik — DISMISSED** [review]

File: services/src/main/java/org/keycloak/services/managers/CodeGenerateUtil.java [hunk 1]
```diff
@@ -173,5 +173,47 @@ public String getClientNote(AuthenticationSessionModel clientSession, String not
         }
     }
 
+    /**
+     * A {@link KeycloakTransaction} that runs a task only when {@link #rollback()} is called.
+     */
+    static class RollbackDrivenTransaction implements KeycloakTransaction {
+
+        private KeycloakSessionFactory factory;
+        private KeycloakSessionTask task;
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/services/managers/CodeGenerateUtil.java:182]: Please make the fields `final`

File: services/src/main/java/org/keycloak/services/managers/CodeGenerateUtil.java [hunk 2]
```diff
@@ -173,5 +173,47 @@ public String getClientNote(AuthenticationSessionModel clientSession, String not
         }
     }
 
+    /**
+     * A {@link KeycloakTransaction} that runs a task only when {@link #rollback()} is called.
+     */
+    static class RollbackDrivenTransaction implements KeycloakTransaction {
```
Comment (hmlnarik) [inline services/src/main/java/org/keycloak/services/managers/CodeGenerateUtil.java:179]: Should this be a `private` class

**Review by hmlnarik — COMMENTED** [review]
Thank you for the fix! Could you please check the inline comments?

**Comment by vramik** [comment]
@sguilhen what is the plan about re-enabling optimistic locking for jpa auth sessions implementation?

**Comment by sguilhen** [comment]
@vramik I thought you would revisit that later, once this PR is merged.

File: services/src/main/java/org/keycloak/services/managers/CodeGenerateUtil.java (hunk 2 above)
Comment (sguilhen) [reply services/src/main/java/org/keycloak/services/managers/CodeGenerateUtil.java]: Absolutely. Changed.

**Comment by vramik** [comment]
> @vramik I thought you would revisit that later, once this PR is merged.

ok, I've created an issue for it: https://github.com/keycloak/keycloak/issues/12242

**Review by hmlnarik — APPROVED** [review]

**Pull request merged**